wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivityManager.java

```diff
@@ -167,12 +171,82 @@ protected long getIdleTimeout(String wsId) {
       delayParameterName = "che.workspace.activity_check_scheduler_period_s")
   private void invalidate() {
     try {
-      activityDao.findExpired(System.currentTimeMillis()).forEach(this::stopExpired);
+      stopAllExpired();
+      checkValidExpiration();
     } catch (ServerException e) {
       LOG.error(e.getLocalizedMessage(), e);
     }
   }
 
+  private void stopAllExpired() throws ServerException {
+    activityDao.findExpired(System.currentTimeMillis()).forEach(this::stopExpired);
+  }
+
+  private void checkValidExpiration() throws ServerException {
+    for (String runningWsId : workspaceRuntimes.getRunning()) {
+      WorkspaceActivity activity = activityDao.findActivity(runningWsId);
+      if (activity == null) {
+        LOG.warn(
+            "Found a running workspace {} without any activity record. This shouldn't really"
+                + " happen but is being rectified by adding a new activity record for it.",
+            runningWsId);
+
+        try {
+          Workspace workspace = workspaceManager.getWorkspace(runningWsId);
+          long createdTime =
+              Long.parseLong(workspace.getAttributes().get(Constants.CREATED_ATTRIBUTE_NAME));
+
+          activity = new WorkspaceActivity();
+          activity.setWorkspaceId(runningWsId);
+          activity.setCreated(createdTime);
+          activity.setStatus(WorkspaceStatus.RUNNING);
+          activity.setLastRunning(System.currentTimeMillis());
+
+          long idleTimeout = getIdleTimeout(runningWsId);
+          if (idleTimeout > 0) {
+            // only set the expiration if it is used...
+            activity.setExpiration(System.currentTimeMillis() + idleTimeout);
+          }
+
+          activityDao.createActivity(activity);
+        } catch (NotFoundException e) {
+          LOG.error(
+              "Detected a running workspace {} but could not find its record.", runningWsId, e);
+        } catch (ConflictException e) {
+          LOG.debug(
+              "Activity record created while we were trying to rectify its absence for a"
+                  + " running workspace {}.",
+              runningWsId);
+        }
+      } else {
+        long idleTimeout = getIdleTimeout(runningWsId);
+        if (idleTimeout == 0) {
+          // expiry not used at all...
```

Comment:
Why do you use `...`? Is it an incomplete sentence or you want to keep the intrigue?

Reply:
It's just me - never want to make definitive statements ;)